jupyterhub/utils.py

```diff
@@ -296,6 +299,30 @@ def metrics_authentication(self):
         raise web.HTTPError(403)
 
 
+def needs_scope(scope):
+    """Decorator to restrict access to users or services with the required scope"""
+
+    def scope_decorator(func):
+        @functools.wraps(func)
+        def _auth_func(self, *args, **kwargs):
+            allows_subset = 'subset' in func.__code__.co_varnames
```

Comment:
we can use [inspect.signature](https://docs.python.org/3/library/inspect.html#inspect.signature) here